src/utils.rs

```diff
@@ -1,33 +1,63 @@
 use crate::AppSW;
 use core::str::from_utf8;
 
-pub const MAX_ALLOWED_PATH_LEN: usize = 10;
+/// BIP32 path stored as an array of [`u32`].
+///
+/// # Generic arguments
+///
+/// * `S` - Maximum possible path length, i.e. the capacity of the internal buffer.
+pub struct Bip32Path<const S: usize = 10> {
+    buffer: [u32; S],
+    len: usize,
+}
 
-/// Convert serialized derivation path to u32 array elements
-pub fn read_bip32_path(data: &[u8], path: &mut [u32]) -> Result<usize, AppSW> {
-    // Check input length and path buffer capacity
-    if data.is_empty() || path.len() < data.len() / 4 {
-        return Err(AppSW::WrongApduLength);
+impl AsRef<[u32]> for Bip32Path {
+    fn as_ref(&self) -> &[u32] {
+        &self.buffer[..self.len]
     }
+}
 
-    let path_len = data[0] as usize; // First byte is the length of the path
-    let path_data = &data[1..];
-
-    // Check path data length and alignment
-    if path_data.len() != path_len * 4
-        || path_data.len() > MAX_ALLOWED_PATH_LEN * 4
-        || path_data.len() % 4 != 0
-    {
-        return Err(AppSW::WrongApduLength);
+impl<const S: usize> Default for Bip32Path<S> {
+    fn default() -> Self {
+        Self {
+            buffer: [0u32; S],
+            len: 0,
+        }
     }
+}
 
-    let mut idx = 0;
-    for (i, chunk) in path_data.chunks(4).enumerate() {
-        path[idx] = u32::from_be_bytes(chunk.try_into().unwrap());
-        idx = i + 1;
-    }
+impl<const S: usize> TryFrom<&[u8]> for Bip32Path<S> {
+    type Error = AppSW;
 
-    Ok(idx)
+    /// Constructs a [`Bip32Path`] from a given byte array.
+    ///
+    /// This method will return an error in the following cases:
+    /// - the input array is empty,
+    /// - the number of bytes in the input array is not a multiple of 4,
+    /// - the input array exceeds the capacity of the [`Bip32Path`] internal buffer.
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - Encoded BIP32 path. First byte is the length of the path, as encoded by ragger.
+    fn try_from(data: &[u8]) -> Result<Self, Self::Error> {
+        // Check data length
+        if data.is_empty() // At least the length byte is required
+            || ((data.len() - 1) / 4 > S)
+            || (data[0] as usize * 4 != data.len() - 1)
+        {
+            return Err(AppSW::WrongApduLength);
+        }
+
+        let mut path = [0; S];
+        for (chunk, p) in data[1..].chunks(4).zip(path.iter_mut()) {
+            *p = u32::from_be_bytes(chunk.try_into().unwrap());
+        }
+
+        Ok(Self {
+            buffer: path,
+            len: (data.len() - 1) / 4,
```

Comment:
Maybe define a local variable `let input_path_len = (data.len() - 1) / 4` that can also be used on line 45 for the checks.

Reply:
Good suggestion.
I updated the code.